app/src/main/java/org/onionshare/android/ui/MainViewModel.kt

```diff
@@ -0,0 +1,79 @@
+package org.onionshare.android.ui
+
+import android.app.Application
+import android.net.Uri
+import android.text.format.Formatter
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import org.onionshare.android.files.FileManager
+import org.onionshare.android.server.SendPage
+import org.onionshare.android.server.WebserverManager
+import org.slf4j.LoggerFactory
+import javax.inject.Inject
+
+private val LOG = LoggerFactory.getLogger(MainViewModel::class.java)
+
+@HiltViewModel
+class MainViewModel @Inject constructor(
+    private val app: Application,
+    handle: SavedStateHandle,
+    private val webserverManager: WebserverManager,
+    private val fileManager: FileManager,
+) : AndroidViewModel(app) {
+
+    private val _fileManagerState = MutableStateFlow<FileManager.State>(FileManager.State.NoFiles)
+    val fileManagerState: StateFlow<FileManager.State> = _fileManagerState
+    val webserverState = webserverManager.state
+
+    override fun onCleared() {
+        super.onCleared()
+        // TODO later, we shouldn't stop the server when closing the activity
+        stopServer()
+    }
+
+    fun onUrisReceived(uris: List<Uri>) {
+        if (uris.isEmpty()) return // user backed out of select activity
+
+        // not supporting selecting entire folders with sub-folders
+        viewModelScope.launch(Dispatchers.IO) {
+            val filesAdded = fileManager.addFiles(uris, fileManagerState.value)
+            _fileManagerState.value = filesAdded
+        }
+    }
+
+    fun startServer() {
+        // FIXME this is a mixing of concerns to get the right state in the UI
+        webserverManager.onFilesBeingZipped()
+
+        viewModelScope.launch(Dispatchers.IO) {
+            val filesAdded = fileManagerState.value as FileManager.State.FilesAdded
```

Comment:
Is it possible that the state has changed before this runs on an IO thread?

Reply:
Yeah, that is the fun part about using new stuff like coroutines here.

We are only ever setting this on the I/O thread, but I think there might be multiple being used.
Not sure if the value can be outdated, but so far in the simple UI, I think it can't go back from `FilesAdded` to `NoFiles`. Maybe the state we are getting here is outdated and still `NoFiles`? Unlikely, but maybe possible.

The setter for the value has this in the docs:

> This property is thread-safe and can be safely updated from concurrent coroutines without external synchronization.